streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1946,6 +1946,17 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
       _ <- pull.toManaged_
     } yield ()
 
+  /**
+   * Locks the execution of this stream to the specified executor. Any streams
+   * that are composed after this one will automatically be shifted back to the
+   * previous executor.
+   */
+  def lock(executor: Executor): ZStream[R, E, O] =
+    ZStream.fromEffect(ZIO.descriptor).flatMap { descriptor =>
+      ZStream.managed(ZManaged.lock(executor)) *>
+        self <* ZStream.fromEffect(ZIO.shift(descriptor.executor))
```

Comment:
Shouldn't we use some type of bracket here so in case there are failures, we shift back?

Reply:
The `ZManaged.lock` is already going to shift back if there is a failure, since that is a managed resource that describes shifting onto the specified executor and shifting back. What we need the `ZIO.shift` for is to make sure that if this stream is successful any continuation is shifted back to the original executor.

Reply:
All right, makes sense!